clustermesh/clustermesh.go

```diff
@@ -2158,3 +2267,64 @@ func updateClustermeshConfig(
 
 	return newValues, nil
 }
+
+func removeFromClustermeshConfig(values map[string]interface{}, clusterName string) (map[string]any, error) {
+	// get current clusters config slice, if it exists
+	c, found, err := unstructured.NestedFieldCopy(values, "clustermesh", "config", "clusters")
+	if err != nil {
+		return nil, fmt.Errorf("existing clustermesh.config is invalid")
+	}
+	if found && c == nil {
+		return nil, fmt.Errorf("active cluster is not connected to any other clusters")
+	}
+	if !found {
+		c = []any{}
+	}
+
+	cs, ok := c.([]any)
+	if !ok {
+		return nil, fmt.Errorf("existing clustermesh.config.clusters array is invalid")
+	}
+	oldClusters := make([]map[string]any, 0, len(cs))
+	for _, m := range cs {
+		cluster, ok := m.(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf("existing clustermesh.config.clusters map is invalid")
+		}
+		name, ok := cluster["name"].(string)
+		if ok && name == clusterName {
+			victim = true
+			continue
+		}
+		return nil, fmt.Errorf("existing clustermesh.config.clusters map is invalid")
+	}
+
+	clusters, victim := map[string]map[string]any{}, false
+	for _, c := range oldClusters {
+		if name, ok := c["name"].(string); ok {
+			if name == aiRemote.ClusterName {
+				victim = true
+				continue
+			}
+			clusters[name] = c
+		}
+	}
+	if !victim {
+		return nil, fmt.Errorf("active cluster is not currently connected to cluster %q", clusterName)
```

Comment:
Should we return an error in case the target cluster is not present, or just ignore it? Let's say that we executed `cilium clustermesh disconnect` and something went wrong while applying the helm changes to the second cluster. At that point we would have a broken setup in which it is impossible to execute disconnect again. 

Another related aspect is whether connect and disconnect changes should be atomic, in the sense that if a failure occurs while applying them to the second cluster we should trigger a helm rollback on the first one. I honestly don't have a strong opinion here. WDYT? (/cc @michi-covalent @asauber)

Reply:
Interesting Q Marco! I do think the connect/disconnect operations should be transactional and not leave the clusters in a dangling state given the symmetrical nature of these calls. 

That said this error does not trigger state change on the clusters as the actual helm update takes place in subsequent stage. But it does leave users with the inability to perform an operation if the validation fails.

Reply:
> That said this error does not trigger state change on the clusters as the actual helm update takes place in subsequent stage. But it does leave users with the inability to perform an operation if the validation fails.

Yeah, it would cause a failure only if the previous attempt failed in the middle of the upgrade. That's something that could be addressed in a followup PR though (probably just moving to a transactional approach).

Reply:
yeah that's probably out of scope for this PR. let me open a github issue to follow up on this 📝

Reply:
filed https://github.com/cilium/cilium-cli/issues/1748 ✅ 😇